### PR TITLE
Overhaul of string/identifier/variable escaping

### LINKED ISSUE #308: More than one string with trailing backslash
**Describe the bug**

I tried formatting

```sql
select
  *
from
  mytable
where
  mycol_a like 'prefix\_%' escape '\'
  and mycol_b like 'prefix\_%' escape '\'
```

**Expected behavior**

Output exactly the same as input.

**Actual behavior**

```
Error: Parse error: Unexpected "\_%'"
    at E.value (https://unpkg.com/sql-formatter@latest/dist/sql-formatter.min.js:1:34317)
    at E.value (https://unpkg.com/sql-formatter@latest/dist/sql-formatter.min.js:1:41347)
    at I.value (https://unpkg.com/sql-formatter@latest/dist/sql-formatter.min.js:1:30283)
    at Module.SS (https://unpkg.com/sql-formatter@latest/dist/sql-formatter.min.js:1:210009)
    at HTMLTextAreaElement.format (https://sql-formatter-org.github.io/sql-formatter/index.js:50:31)
```

If I add a space after the first `'\'` to `'\ '` it works. If I remove the last `and [etc.]` line then it also works.

**Usage**

- How are you calling / using the library? https://sql-formatter-org.github.io/sql-formatter/
- What SQL language(s) does this apply to? I tried every formatter. This is definitely [valid syntax in Trino](https://github.com/trinodb/trino/blob/91ce0ec5c98a5019f5fb66683bae0ee6c18201b0/core/trino-parser/src/main/antlr4/io/trino/sql/parser/SqlBase.g4#L1147) but also other SQL dialects I think
- Which SQL Formatter version are you using? `latest`. v4.0.0 (with some modifications which shouldn't affect this case) adds a bunch of spaces to the last line `and mycol_b like ' prefix \ _ % ' escape ' \ '`

### PR DESCRIPTION
**Pull request by nene**
instead of quote types `""` and `''`, there are now four versions of each to pick the escaping style:

- `""-qq` - with repeated-quote escape
- `""-bs` - with backslash escape
- `""-qq-bs` - with both repeated-quote and backslash escape
- `""-raw` - with out any escaping

The old plain `""` and `''` have been dropped, to force one to explicitly make a choice of escaping style.

Added bunch of additional tests to check for escaping (and also to check for particular styles of escaping not working).

Also dropped support for incomplete strings (like `"...`), just like I previously dropped support for incomplete block-comments.

Fixes #308

### CODE REVIEW

File: src/lexer/regexFactory.ts [hunk 1]
```diff
@@ -96,25 +96,35 @@ const buildQStringPatterns = () => {
...
+  // PostgreSQL dollar-quoted
+  '$$': String.raw`(?<tag>\$\w*\$)[\s\S]*?\k<tag>`,
+  // BigQuery '''triple-quoted''' (using \' to escape)
+  "'''..'''": String.raw`'''[^\\]*?(?:\\.[^\\]*?)*?'''`,
+  // BigQuery """triple-quoted""" (using \" to escape)
+  '""".."""': String.raw`"""[^\\]*?(?:\\.[^\\]*?)*?"""`,
+  // Hive and Spark variables: ${name}
+  '{}': String.raw`(?:\{[^\}]*\})`,
+  // Oracle q'' strings: q'<text>' q'|text|' ...
   "q''": buildQStringPatterns(),
```
Comment (nene) [inline src/lexer/regexFactory.ts:128]: The main changes here involve `""` and `''`.

The only change done with all other quote styles is that support for missing closing quote has been dropped.

File: test/features/identifiers.ts [hunk 2]
```diff
@@ -71,15 +80,22 @@ export default function supportsIdentifiers(format: FormatFn, identifierTypes: I
       `);
     });
 
+    it('supports escaping in U&"" strings by repeated quote', () => {
+      expect(format('U&"foo "" JOIN bar"')).toBe('U&"foo "" JOIN bar"');
+    });
+
     it('detects consecuitive U&"" identifiers as separate ones', () => {
```
Comment (inferrinizzard) [inline test/features/identifiers.ts:87]:
```suggestion
    it('detects consecutive U&"" identifiers as separate ones', () => {
```

**Comment by inferrinizzard** [comment]
would this -qq and -bs suffix also work for some of the prefixed quote types ?
I know a few of the dialect have similar prefixed- types but may have different escape methods between them

**Review by inferrinizzard — APPROVED** [review]

**Comment by nene** [comment]
> would this -qq and -bs suffix also work for some of the prefixed quote types ?

Yep. I'm already using it for various prefixed strings. For example the `-raw` suffix with `R'...'` strings.

**Pull request merged**